### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized attention/xgbst predictor model with integer arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#350]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(24957580949472).epochs(1).int(1).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold. The earliest failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (`recipe.rs` ~2828-2896) on the NVIDIA `ptx`/`ptxas` branch before any epoch arithmetic. The model uses quantized attention and predictor blocks:
`.attn(1).ln().norm(batch).qi(5).1`
`.xgbst().huber().norm(batch).qi(4).nf`
`.loss(bce)`
The loader (`data/temporal/chronological_splits`) and `.int(1)` arithmetic execute after the native compilation hang. No existing nv0 stall issue covers this attention + xgbst predictor quantization path (308 = pool/attention, 271 = residual, 324 = residual/cbst, 317 = pool/cbst). Expected sub-second epoch for this small temporal load; 60s is materially slower and caused by the NVIDIA native compiler hang, not workload size or memory traffic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `e85665c92191fde2`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized attention/xgbst predictor model with integer arithmetic",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.001).seed(24957580949472).epochs(1).int(1).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold. The earliest failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` (`recipe.rs` ~2828-2896) on the NVIDIA `ptx`/`ptxas` branch before any epoch arithmetic. The model uses quantized attention and predictor blocks:\n`.attn(1).ln().norm(batch).qi(5).1`\n`.xgbst().huber().norm(batch).qi(4).nf`\n`.loss(bce)`\nThe loader (`data/temporal/chronological_splits`) and `.int(1)` arithmetic execute after the native compilation hang. No existing nv0 stall issue covers this attention + xgbst predictor quantization path (308 = pool/attention, 271 = residual, 324 = residual/cbst, 317 = pool/cbst). Expected sub-second epoch for this small temporal load; 60s is materially slower and caused by the NVIDIA native compiler hang, not workload size or memory traffic.",
  "rationale": "Performance packet with `kind=performance`, `stable:true`, `elapsed_seconds:60`, `backend:nv0`. Inspecting source (`recipe.rs` 2868-2913) shows the hang reaches `compile_native_artifact` before any training arithmetic. The reproduction model (`.attn(1).ln().norm(batch).qi(5).1` then `.xgbst().huber().norm(batch).qi(4).nf` with `.int(1)`) triggers native artifact compilation through quantized attention and predictor nodes, a block combination not represented by any open nv0 stall issue (271/308/316/319/324/317/291/348/345/349/322/302/314/321/325/347). The loader, seed, `.int(1)`, `.loss(bce)`, and precision settings execute after the stall and do not change the earliest cause. For this workload the arithmetic and memory cost is minimal; 60s versus expected <1s confirms the dominant gap is NVIDIA native compiler hang."
}
```

## Deterministic failure packet

kind=performance
id=e85665c92191fde2
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:9888 next:9889 composition:24957580949472
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-9888-5.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target");
	let model = recipe.model()
		.attn(1).ln().norm(batch).qi(5).1
		.xgbst().huber().qi(4).nf
		.loss(bce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(24957580949472)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.int(1)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Performance stall on backend nv0: `recipe.train().run()` remained active past 60 s. The model starts with `.xgbst().sigmoid().iq(4).xs .forest(2).huber().iq(3).m`, which lowers to a `Primitive::Predictor` (XGBoost) whose execution loop runs in the compiled native artifact. Replay is stable, so the delay is an execution stall rather than a one‑time compilation delay.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `90f768cf9880c7e6`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"Performance stall on backend nv0: `recipe.train().run()` remained active past 60 s. The model starts with `.xgbst().sigmoid().iq(4).xs .forest(2).huber().iq(3).m`, which lowers to a `Primitive::Predictor` (XGBoost) whose execution loop runs in the compiled native artifact. Replay is stable, so the delay is an execution stall rather than a one‑time compilation delay.","issue":350,"rationale":"The packet’s earliest public failure is the training epoch crossing the 60‑second slow‑runtime threshold on backend nv0. Inspection of `recipe.rs` shows the XGBST block creates a `Primitive::Predictor` that is emitted via `emit_predictor_forward` and executed in the compiled native artifact. Issue #350 already documents this XGBST predictor stall on nv0; the current configuration differences (e.g., quantization variants) do not alter the earliest failing operation or implementation path, so this packet is additional evidence for the existing issue.","title":"NVIDIA nv0 training stalls beyond 60s on quantized attention/xgbst predictor model with integer arithmetic","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=90f768cf9880c7e6
base=commit=c6a63621f4e7857dd2e29d5ea99598229635dc71 tracked_tree=clean
cursor=seed:0 cursor:15305 next:15306 composition:108139687881355
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-22-15305/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.xgbst().sigmoid().iq(4).xs
		.forest(2).huber().iq(3).m
		.loss(ce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(108139687881355)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.int(8)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #439.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #439: Verify current NVIDIA attention predictor execution</summary>

Closes #350

## Verification

- Ran the reported temporal split data, Attention and XGBoost model, quantization, INT1 arithmetic, optimizer, seed, stopping threshold, and one-epoch training path through `./target/release/recipe --device benji:nv0`. Only the saved bundle path changed from `/tmp` to `/home/nate/.cache/recipe/issues/350/bundle.ogdl` because `/tmp` is tmpfs.
- The current loader discarded trailing blank records and produced 18,000 rows with an input width of 20.
- The Attention prefix, XGBoost surrogate, and main training native artifacts were cache misses. Each completed Clang LLVM bitcode generation and `llc` PTX generation. The separate 64-row calibration artifact was a cache hit.
- The process exited 0 in 10.97 seconds. The epoch took 198.486 milliseconds, BCE was finite at 0.6931, and the saved bundle was 383,424 bytes.
- The reported native-artifact compile stall does not reproduce on current minimal.

## Change

This verification commit changes no tracked files because current behavior does not justify a production-code change.

</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public `recipe.train().run()` remains active beyond 60 seconds on `nv0`. The earliest failing operation is native artifact compilation through `compile_model` -> `compile_native_artifact` -> NVIDIA LLVM/PTXAS for the quantized attention and XGBoost predictor graph. With a small image sample table, one attention pass, and XGBoost's 8 boosting iterations at depth 5, the expected runtime is seconds rather than 60+ seconds. The dominant gap is the NVIDIA native compilation stall, providing additional evidence for this issue.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `78af96c85c9f1099`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 350,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized attention/xgbst predictor model with integer arithmetic",
  "body": "The public `recipe.train().run()` remains active beyond 60 seconds on `nv0`. The earliest failing operation is native artifact compilation through `compile_model` -> `compile_native_artifact` -> NVIDIA LLVM/PTXAS for the quantized attention and XGBoost predictor graph. With a small image sample table, one attention pass, and XGBoost's 8 boosting iterations at depth 5, the expected runtime is seconds rather than 60+ seconds. The dominant gap is the NVIDIA native compilation stall, providing additional evidence for this issue.",
  "rationale": "Issue #350 represents the same earliest public failure and implementation path: a quantized attention plus XGBoost predictor composition reaches NVIDIA native artifact compilation and crosses the stable 60-second threshold. The `.qi(5).k.m` and `.qi(3).k` variants, `.int(4)` arithmetic, image multiview dataset, seed, and loss do not establish a different defect. The inspected source confirms attention lowering through `lower_attention`, XGBoost lowering through `lower_estimator` and `fit_xgboost`, and the shared NVIDIA LLVM/PTXAS path in `compile_native_artifact`. The workload is materially too small to justify the observed runtime."
}
```

## Deterministic failure packet

kind=performance
id=78af96c85c9f1099
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:23598 next:23599 composition:96632063639082
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-7987-23598/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/image/sample_subfolders_multiview")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.attn(1).gelu().norm(batch).qi(5).k.m
		.xgbst().relu().norm(batch).qi(3).k
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(96632063639082)
		.epochs(1)
		.log(all)
		.int(4)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0
